### PR TITLE
AP_DroneCAN: add support for detecting downed link

### PR DESCRIPTION
**Pull request by bugobliterator**
* Follows similar implementation down for AP_Periph https://github.com/ArduPilot/ardupilot/pull/24840/commits/d892df492bebba23ebf42186f3179533891dc7cb but for DroneCAN. This ensures that we are not filling up buffers with stale transactions just because second interface is down/disconnected.

### CODE REVIEW

File: libraries/AP_DroneCAN/AP_Canard_iface.cpp [hunk 1]
```diff
@@ -96,7 +96,8 @@ bool CanardInterface::broadcast(const Canard::Transfer &bcast_transfer) {
 #endif
     };
     // do canard broadcast
-    int16_t ret = canardBroadcastObj(&canard, &tx_transfer);
+    volatile int16_t ret;
```
Comment (tridge) [inline libraries/AP_DroneCAN/AP_Canard_iface.cpp:99]: why volatile?

File: libraries/AP_DroneCAN/AP_Canard_iface.cpp [hunk 2]
```diff
@@ -214,6 +215,18 @@ void CanardInterface::processTx(bool raw_commands_only = false) {
...
         }
+        const auto *stats = ifaces[iface]->get_statistics();
+        bool iface_down = true;
+        if (stats == nullptr || (AP_HAL::micros64() - stats->last_transmit_us) < 200000UL) {
+            /*
+            We were not able to queue the frame for
+            sending. Only mark the send as failing if the
+            interface is active. We consider an interface as
+            active if it has had a successful transmit in the
+            last 500 milliseconds
```
Comment (tridge) [inline libraries/AP_DroneCAN/AP_Canard_iface.cpp:226]: 200ms

**Review by tridge — CHANGES_REQUESTED** [review]
minor fixup for volatile

File: libraries/AP_DroneCAN/AP_Canard_iface.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_DroneCAN/AP_Canard_iface.cpp]: fixed

File: libraries/AP_DroneCAN/AP_Canard_iface.cpp [hunk 3]
```diff
@@ -214,6 +214,17 @@ void CanardInterface::processTx(bool raw_commands_only = false) {
         if (txq == nullptr) {
             return;
         }
+        const auto *stats = ifaces[iface]->get_statistics();
+        bool iface_down = true;
+        if (stats == nullptr || (AP_HAL::micros64() - stats->last_transmit_us) < 200000UL) {
```
Comment (tridge) [inline libraries/AP_DroneCAN/AP_Canard_iface.cpp:219]: i think we need to fetch the last_transmit_us into a variable first, then get the micros64() into a variable, then subtract
the issue is that if the CAN thread updates last_transmit_us after we do the micros64() call then we will get a very large number, and will think the interface is down, so may lose a packet

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_DroneCAN/AP_Canard_iface.cpp (hunk 3 above)
Comment (bugobliterator) [reply libraries/AP_DroneCAN/AP_Canard_iface.cpp:219]: Correct, I also marked *stats as volatile so compiler doesn't try to optimise it.

Comment (bugobliterator) [reply libraries/AP_DroneCAN/AP_Canard_iface.cpp]: Also the issue was present in the AP_Periph implementation as well, so fixed there as well.

**Review by tridge — APPROVED** [review]

**Pull request merged**